pysmt/smtlib/script.py

```diff
@@ -255,14 +259,53 @@ def get_last_formula(self, mgr=None):
             if cmd.name == smtcmd.RESET_ASSERTIONS:
                 stack = []
                 backtrack = []
+                goals = []
+                goals_backtrack = []
+                max_smt_goals = {}
+                max_smt_goals_backtrack = defaultdict(list)
             elif cmd.name == smtcmd.PUSH:
                 for _ in range(cmd.args[0]):
                     backtrack.append(len(stack))
+                    goals_backtrack.append(len(goals))
+                    for k, (_, goal) in max_smt_goals.items():
+                        max_smt_goals_backtrack[k].append(len(goal.soft))
             elif cmd.name == smtcmd.POP:
                 for _ in range(cmd.args[0]):
                     l = backtrack.pop()
                     stack = stack[:l]
-
+                    l = goals_backtrack.pop()
+                    goals = goals[:l]
+                    # a max_smt_goal might be removed from the goals completely with the pop
+                    # so we remove it from the dict, otherwise we can just remove the
+                    # last l elements from the list of formulae
+                    goals_to_remove = []
+                    for k, (goal_position, goal) in max_smt_goals.items():
+                        if goal_position >= len(goals):
+                            goals_to_remove.append(k)
+                        else:
+                            l = max_smt_goals_backtrack[k].pop()
+                            goal.soft = goal.soft[:l]
+                    for k in goals_to_remove:
+                        del max_smt_goals[k]
+                        del max_smt_goals_backtrack[k]
+            elif cmd.name in (
+                smtcmd.MAXIMIZE,
+                smtcmd.MINIMIZE,
+                smtcmd.MAXMIN,
+                smtcmd.MINMAX
+            ):
+                goals.append(_parse_goal(cmd))
+            elif cmd.name == smtcmd.ASSERT_SOFT:
```

Comment:
why not merging this case with the previous one? Not handling the `None` case is not significant from the performances point of view and makes some assumptions on the behavior of the `_parse_goal` function...

Reply:
Also, even if you pass the extra arguments should not be a problem IMHO

Reply:
Also, looking at the function below, maybe it is simpler to inline it here since you split on all the cases anyway...